data_analyzer.py

```diff
@@ -113,3 +127,108 @@ def init_dataframe_differences():
+
+def create_vector(point_A, point_B):
+    return [point_B.LAT - point_A.LAT, point_B.LON - point_A.LON]
+
+def shortest_distance(x1, y1, a, b, c):  
+    perpendicular = abs((a * x1 + b * y1 + c)) / (math.sqrt(a ** 2 + b ** 2)) 
+    return perpendicular
+
+def find_shortest_distance(point1, point2, point3):
+        x = [point2.LAT, point3.LAT]
+        y = [point2.LON, point3.LON]
+        slope , intercept,_,_,_ = stats.linregress(x, y)
+        return shortest_distance(point1.LAT, point1.LON, slope, -1, intercept)
+
+def lap1_generator():
+    for file_name in listdir('data/laps'):
+        if not file_name.startswith('lap1'): continue
+        yield init_dataframe('data/laps/'+file_name)
+
+def lap2_generator():
+    for file_name in listdir('data/laps'):
+        if file_name.startswith('lap1'): continue
+        yield init_dataframe('data/laps/'+file_name)
+        
+def find_out_difference_v2(lap, ref_lap):
+    sum_of_distances = 0
+    for i in lap.index:
+        point = lap.loc[i]
+
+        closest_index = find_closest_point(point, ref_lap)
+        closest_point = ref_lap.loc[closest_index]
+
+        neighbor_i = len(ref_lap)-1 if closest_index == 0 else closest_index-1
+        neighbor1 = ref_lap.loc[neighbor_i]
+        neighbor_i = 0 if len(ref_lap) == closest_index+1 else closest_index+1
+        neighbor2 = ref_lap.loc[neighbor_i]
+
+        v1 = create_vector(closest_point, point)
+        v2 = create_vector(closest_point, neighbor1)
+        v3 = create_vector(closest_point, neighbor2)
+
+        angle1 = find_angle_between_vectors(v1,v2)
+        angle2 = find_angle_between_vectors(v1,v3)
+
+        degrees90 = math.pi / 2
+        min_dist = -1
+        if angle1 > degrees90 and angle2 > degrees90:
+            min_dist = line_length(point.LAT, point.LON, closest_point.LAT, closest_point.LON)
+        elif angle1 < degrees90 and angle2 < degrees90:
+            dist1 = find_shortest_distance(point, closest_point, neighbor1)
+            dist2 = find_shortest_distance(point, closest_point, neighbor2)
+            min_dist = dist1 if dist1 <= dist2 else dist2
+        elif angle1 <= degrees90:
+            min_dist = find_shortest_distance(point, closest_point, neighbor1)
+        elif angle2 <= degrees90:
+            min_dist = find_shortest_distance(point, closest_point, neighbor2)
+
+        if min_dist == -1:
+            print ('ERROR: Could not find distance')       
+            print("Indices: {} {}\nAngles: {} {}".format(i, closest_index, angle1, angle2))
+        elif math.isnan(min_dist):
+            print("NAN value!!!\nIndices: {} {}\nAngles: {} {}".format(i, closest_index, angle1, angle2))
+        elif min_dist < 0:
+            print("Negative value!!!\nIndices: {} {}\nAngles: {} {}".format(i, closest_index, angle1, angle2))
+        else:
+            min_dist = degrees2kilometers(min_dist) * 100000    # in centimeters
+            sum_of_distances += min_dist        
```

Comment:
Instead of writing all these functions you could have just used packages sympy and scipy.spatial (both used in data_analyzer.ipynb), maybe consider refactoring and using those packages, the code is not that easy to understand this way

Reply:
I was trying to find libraries for this purpose, but no luck. I should have asked you about that before =)
But I think I will not refactor it because there is a big chance that we won't use this code anymore. 
I'll do it if it is used somehow in our project.